core/src/test/java/io/forestframework/core/http/HandlerIntegrationTest.java

```diff
@@ -0,0 +1,458 @@
+
+    @Route(value = "/**", type = RoutingType.HANDLER)
+    public void handler(HttpServerResponse response) {
+        runUnderTrace("shouldNotContinueInHandler");
+    }
+
+    @OnError(value = "/**")
+    public void customErrorHandler(HttpServerResponse response, Throwable e) {
+        response.write(" should be handled by custom error handler. ");
+        response.write(" e.getMessage(): " + e.getMessage());
+        runUnderTrace("shouldBeHandledByCustomErrorHandler");
+    }
+
+    @PostHandler("/**")
+    public void postHandler() {
+        runUnderTrace("shouldContinueInPostHandler");
+    }
+}
+
+@Router("/errorInPostHandler")
+class ErrorInPostHandler extends AbstractHandlerIntegrationTest {
+
+    @Get("/**")
+    public void handler(HttpServerResponse response) {
+        runUnderTrace("handler");
+    }
+
+    @OnError(value = "/**")
+    public void customErrorHandler(HttpServerResponse response, Throwable e) {
+        response.write(" should not be handled by custom error handler. ");
+        response.write(" e.getMessage(): " + e.getMessage());
+        runUnderTrace("shouldNotBeHandledByCustomErrorHandler");
+    }
+
+    @PostHandler("/**")
+    public void postHandler(HttpServerResponse response) {
+        runUnderTrace("errorInPostHandler");
+        throw new HttpException(HttpStatusCode.CLIENT_ERROR);
+    }
+}
+
+@Router("/preHandlerNotReturnBoolean")
+class ThrowExceptionsWhenPreHandlersNotReturnBoolean extends AbstractHandlerIntegrationTest {
+
+    @PreHandler("/**")
+    public String preHandler(HttpServerRequest request) {
+        runUnderTrace("prehandlerReturnsString");
+        return "shouldNotReturnString";
+    }
+
+    @Route(value = "/**", type = RoutingType.HANDLER)
+    public void handler(HttpServerRequest request, HttpServerResponse response) {
+        runUnderTrace("handler");
+    }
+
+    @OnError(value = "/**")
+    public void customErrorHandler(HttpServerResponse response, Throwable e) {
+        response.write(" should be handled by custom error handler. ");
+        response.write(" e.getMessage(): " + e.getMessage());
+        runUnderTrace("customErrorHandler");
+    }
+
+    @PostHandler("/**")
+    public void postHandler(HttpServerResponse response) {
+        runUnderTrace("postHandler");
+    }
+}
+
+class AbstractHandlerIntegrationTest {
+    List<String> traces = Collections.synchronizedList(new ArrayList<>());
+
+    void runUnderTrace(String traceId) {
+        traces.add(traceId);
+    }
+}
+
+@ExtendWith(ForestExtension.class)
+@ForestTest(appClass = MyCustomErrorHandlerApp.class)
+@DisableAutoScan
+@IncludeComponents(classes = {
+        MyCustom500HandlerRouter.class,
+        ErrorInCustomErrorHandler2.class,
+        Custom400ErrorHandler.class,
+        Uncaught404Error.class,
+        PrehandlerErrorPreventPropagation.class,
+        ErrorInPostHandler.class,
+        ThrowExceptionsWhenPreHandlersNotReturnBoolean.class,
+        PreHandlerReturnsVoidTrueOrFalse.class
```

Comment:
Can we split these classes to different test classes?

Reply:
Sure